### PR TITLE
[FIX JENKINS-45801] and [FIX JENKINS-49219], when calculate change log of externals, add module location's credentials as default credentials

### PR DESCRIPTION
**Pull request by Liyaa**
when calculate change log of externals add module location's credentials as default credentials, fix [JENKINS-45801] and [JENKINS-49219]

### CODE REVIEW

**Comment by swidbsk** [comment]
This seems to have fixed spurious authentication issues on repos with lots of externals for us

**Comment by samitok** [comment]
This solves our problem with externals directing in the same repository. I also confirm.

File: src/main/java/hudson/scm/SubversionChangeLogBuilder.java [hunk 1]
```diff
@@ -101,7 +101,7 @@ public SubversionChangeLogBuilder(Run<?,?> build, FilePath workspace, @Nonnull S
         this.env = env;
     }
 
-    public boolean run(Collection<SubversionSCM.External> externals, Result changeLog) throws IOException, InterruptedException {
+    public boolean run(Map<String, List<SubversionSCM.External>> externalsMap, Result changeLog) throws IOException, InterruptedException {
```
Comment (kuisathaverat) [inline src/main/java/hudson/scm/SubversionChangeLogBuilder.java]: externalsMap should be @nonnull anotated

File: src/main/java/hudson/scm/SubversionChangeLogBuilder.java [hunk 2]
```diff
@@ -101,7 +101,7 @@ public SubversionChangeLogBuilder(Run<?,?> build, FilePath workspace, @Nonnull S
         this.env = env;
     }
 
-    public boolean run(Collection<SubversionSCM.External> externals, Result changeLog) throws IOException, InterruptedException {
+    public boolean run(@Nonnull Map<String, List<SubversionSCM.External>> externalsMap, Result changeLog) throws IOException, InterruptedException {
```
Comment (dwnusbaum) [inline src/main/java/hudson/scm/SubversionChangeLogBuilder.java]: This introduces a binary incompatibility, but [it doesn't look like any other plugins call this method](https://github.com/search?utf8=%E2%9C%93&q=user%3Ajenkinsci+SubversionChangeLogBuilder&type=Code). Might be good to add `@Restricted(NoExternalUse.class)` to the class in a followup PR.

File: src/main/java/hudson/scm/SubversionChangeLogBuilder.java [hunk 3]
```diff
@@ -122,24 +122,20 @@ public boolean run(Collection<SubversionSCM.External> externals, Result changeLo
...
                 manager.dispose();
             }
         }
-        ISVNAuthenticationProvider authProvider =
-                CredentialsSVNAuthenticationProviderImpl
-                        .createAuthenticationProvider(build.getParent(), scm, null, listener);
-        final SVNClientManager manager = SubversionSCM.createClientManager(authProvider).getCore();
-        try {
-            SVNLogClient svnlc = manager.getLogClient();
-            for(SubversionSCM.External ext : externals) {
```
Comment (dwnusbaum) [inline src/main/java/hudson/scm/SubversionChangeLogBuilder.java]: Is there any case in which this is still necessary as a fallback?

File: src/main/java/hudson/scm/SubversionChangeLogBuilder.java (hunk 2 above)
Comment (kuisathaverat) [reply src/main/java/hudson/scm/SubversionChangeLogBuilder.java]: I will manage it

File: src/main/java/hudson/scm/SubversionChangeLogBuilder.java (hunk 3 above)
Comment (kuisathaverat) [reply src/main/java/hudson/scm/SubversionChangeLogBuilder.java]: before of this change, there was only one list of remotes now are classified by remote, so now only process external when there are externals, the removed login it is not needed anymore.

Comment (dwnusbaum) [reply src/main/java/hudson/scm/SubversionChangeLogBuilder.java]: Ok, seems fine to me then.

**Comment by stefanbesler** [comment]
Hi!
Lately I am getting loads of

org.tmatesoft.svn.core.SVNCancelException: svn: E200015: E200015: ISVNAuthentication provider did not provide credentials; HTTP authorization 

exceptions (randomly ~ second build) and I hope that this commit solves this issue as I am dealing with externals on the same server too. @kuisathaverat when are you planning to release a new version containing this fix?

**Pull request merged**